go/vt/vttablet/tabletmanager/rpc_external_reparent.go

```diff
@@ -135,24 +132,31 @@ func (agent *ActionAgent) TabletExternallyReparented(ctx context.Context, extern
 }
 
 // finalizeTabletExternallyReparented performs slow, synchronized reconciliation
-// tasks that ensure topology is self-consistent, and then marks the reparent as
-// finished by updating the global shard record.
+// tasks that ensure topology is self-consistent
+// it first updates new and old master tablet records, then updates
+// the global shard record, then refreshes the old master
+// after that it attempts to detect and clean up any lingering old masters
+// note that an up-to-date shard record does not necessarily mean that
+// the reparent completed all the actions successfully
 func (agent *ActionAgent) finalizeTabletExternallyReparented(ctx context.Context, si *topo.ShardInfo, ev *events.Reparent) (err error) {
 	var wg sync.WaitGroup
 	var errs concurrency.AllErrorRecorder
 	var oldMasterTablet *topodatapb.Tablet
 	oldMasterAlias := si.MasterAlias
 
-	// Update the tablet records concurrently.
+	// Update the new and old master tablet records concurrently.
 	event.DispatchUpdate(ev, "updating old and new master tablet records")
 	log.Infof("finalizeTabletExternallyReparented: updating tablet records")
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		// Update our own record to master.
+		log.Infof("finalizeTabletExternallyReparented: updating tablet record for new master: %v", agent.TabletAlias)
+		// Update our own record to master if needed
 		_, err := agent.TopoServer.UpdateTabletFields(ctx, agent.TabletAlias,
 			func(tablet *topodatapb.Tablet) error {
-				tablet.Type = topodatapb.TabletType_MASTER
+				if tablet.Type != topodatapb.TabletType_MASTER {
+					tablet.Type = topodatapb.TabletType_MASTER
```

Comment:
Now that we got rid of the short circuit, maybe we should be more careful here. To actually prevent the update request from being sent (and keep topo traffic as low as possible), we should return `NoUpdateNeeded` if the type is already what we want. Same for updating the old master tablet.